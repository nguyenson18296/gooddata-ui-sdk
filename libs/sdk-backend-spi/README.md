# Analytical Backend Service Provider Interface (SPI)

This package provides definitions of the Service Provider Interface (SPI) for the Analytical Backend.

The interface defines functionality to be implemented for a particular backend to be used in GoodData.UI.

The Analytical Backend SPI for the GoodData platform (codename `bear` in [`@gooddata/sdk-backend-bear`](https://www.npmjs.com/package/@gooddata/sdk-backend-bear)) is fully implemented.

The Analytical Backend SPI for GoodData.CN (codename `tiger` in [`@gooddata/sdk-backend-tiger`](https://www.npmjs.com/package/@gooddata/sdk-backend-tiger)) is almost fully implemented.

## License

(C) 2017-2021 GoodData Corporation

This project is under MIT License. See [LICENSE](https://github.com/gooddata/gooddata-ui-sdk/blob/master/libs/sdk-backend-spi/LICENSE).
