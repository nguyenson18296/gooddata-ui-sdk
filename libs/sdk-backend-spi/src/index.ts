// (C) 2019-2021 GoodData Corporation
/**
 * @packageDocumentation
 */
export {
    IAnalyticalBackend,
    IAnalyticalBackendConfig,
    AnalyticalBackendFactory,
    prepareExecution,
    IAuthenticationContext,
    IAuthenticatedPrincipal,
    IAuthenticationProvider,
    NotAuthenticatedHandler,
} from "./backend";

export { IBackendCapabilities } from "./backend/capabilities";

export {
    ISettings,
    IUserSettings,
    IWorkspaceSettings,
    IUserWorkspaceSettings,
    ISeparators,
} from "./common/settings";

export { IUserService, IUser, userFullName } from "./user";
export { IUserSettingsService } from "./user/settings";

export { IExecutionFactory, IPreparedExecution, IExecutionResult, IDataView } from "./workspace/execution";

export {
    DataValue,
    IMeasureDescriptor,
    IDimensionItemDescriptor,
    IDimensionDescriptor,
    IAttributeDescriptor,
    IMeasureGroupDescriptor,
    IResultAttributeHeader,
    IResultHeader,
    IResultMeasureHeader,
    IResultTotalHeader,
    ITotalDescriptor,
    IResultWarning,
    isAttributeDescriptor,
    isMeasureGroupDescriptor,
    isTotalDescriptor,
    isMeasureDescriptor,
    isResultAttributeHeader,
    isResultMeasureHeader,
    isResultTotalHeader,
    resultHeaderName,
    attributeDescriptorLocalId,
    attributeDescriptorName,
} from "./workspace/execution/results";

export { IWorkspaceSettingsService } from "./workspace/settings";

export {
    IGetVisualizationClassesOptions,
    IWorkspaceInsightsService,
    InsightOrdering,
    IInsightsQueryOptions,
    IInsightsQueryResult,
    IInsightReferences,
    IInsightReferencing,
    InsightReferenceTypes,
    SupportedInsightReferenceTypes,
} from "./workspace/insights";

export {
    IWorkspaceCatalogFactory,
    IWorkspaceCatalogAvailableItemsFactory,
    IWorkspaceCatalog,
    IWorkspaceCatalogFactoryOptions,
    IWorkspaceCatalogWithAvailableItems,
    IWorkspaceCatalogWithAvailableItemsFactoryOptions,
    IWorkspaceCatalogFactoryMethods,
    IWorkspaceCatalogMethods,
} from "./workspace/ldm/catalog";

export { IWorkspaceDatasetsService } from "./workspace/ldm/datasets";

export {
    IElementsQueryFactory,
    IElementsQueryResult,
    IElementsQuery,
    IElementsQueryOptions,
    IElementsQueryAttributeFilter,
} from "./workspace/attributes/elements";

export { IExportConfig, IExportResult } from "./workspace/execution/export";

export {
    IWorkspaceStylingService,
    ThemeFontUri,
    ThemeColor,
    IThemeColorFamily,
    IThemeComplementaryPalette,
    IThemeWidgetTitle,
    IThemeTypography,
    IThemePalette,
    IThemeKpi,
    IThemeChart,
    IThemeTable,
    ITheme,
} from "./workspace/styling";

export {
    AnalyticalBackendError,
    NoDataError,
    DataTooLargeError,
    ProtectedDataError,
    UnexpectedResponseError,
    UnexpectedError,
    NotSupported,
    NotImplemented,
    NotAuthenticated,
    ErrorConverter,
    isAnalyticalBackendError,
    isNoDataError,
    isDataTooLargeError,
    isProtectedDataError,
    isUnexpectedResponseError,
    isUnexpectedError,
    isNotSupported,
    isNotImplemented,
    isNotAuthenticated,
    AuthenticationFlow,
    AnalyticalBackendErrorTypes,
} from "./errors";

export { IPagedResource } from "./common/paging";

export {
    IAnalyticalWorkspace,
    IWorkspacesQuery,
    IWorkspacesQueryFactory,
    IWorkspacesQueryResult,
    IWorkspaceDescriptor,
} from "./workspace";

export {
    IWorkspacePermissionsService,
    IWorkspacePermissions,
    WorkspacePermission,
} from "./workspace/permissions";

export { IWorkspaceAttributesService } from "./workspace/attributes";

export { IWorkspaceMeasuresService } from "./workspace/measures";

export { IWorkspaceFactsService } from "./workspace/facts";

export { IWorkspaceDashboardsService } from "./workspace/dashboards";
export { IDashboardObjectIdentity } from "./workspace/dashboards/common";
export {
    DrillDefinition,
    InsightDrillDefinition,
    KpiDrillDefinition,
    DrillOrigin,
    DrillOriginType,
    DrillTransition,
    DrillType,
    IDrill,
    IDrillFromMeasure,
    IDrillFromAttribute,
    IDrillOrigin,
    IDrillToDashboard,
    IDrillToInsight,
    IDrillToLegacyDashboard,
    isDrillToDashboard,
    isDrillToInsight,
    isDrillToLegacyDashboard,
    IDrillToAttributeUrl,
    isDrillToAttributeUrl,
    IDrillToCustomUrl,
    isDrillToCustomUrl,
    IDrillTarget,
    IDrillToAttributeUrlTarget,
    IDrillToCustomUrlTarget,
    isDrillFromAttribute,
    isDrillFromMeasure,
} from "./workspace/dashboards/drills";
export {
    IDashboard,
    IDashboardReferences,
    IDashboardWithReferences,
    IDashboardDefinition,
    IListedDashboard,
    IDashboardBase,
    IDashboardDateFilterConfig,
    DashboardDateFilterConfigMode,
    IDashboardDateFilterAddedPresets,
} from "./workspace/dashboards/dashboard";
export {
    IFilterContext,
    AbsoluteType,
    DateFilterType,
    FilterContextItem,
    IDashboardAttributeFilterParent,
    IDashboardAttributeFilter,
    IDashboardDateFilter,
    IFilterContextDefinition,
    RelativeType,
    ITempFilterContext,
    isFilterContext,
    isFilterContextDefinition,
    isTempFilterContext,
    isDashboardAttributeFilter,
    isDashboardDateFilter,
    IDashboardFilterReference,
    IDashboardAttributeFilterReference,
    IDashboardDateFilterReference,
    isDashboardAttributeFilterReference,
    isDashboardDateFilterReference,
    IFilterContextBase,
    dashboardFilterReferenceObjRef,
} from "./workspace/dashboards/filterContext";
export {
    IDashboardLayout,
    DashboardWidget,
    IDashboardLayoutSection,
    IDashboardLayoutSectionHeader,
    IDashboardLayoutSize,
    IDashboardLayoutSizeByScreenSize,
    IDashboardLayoutItem,
    ScreenSize,
    isDashboardLayout,
    isDashboardLayoutSection,
    isDashboardLayoutItem,
    isDashboardWidget,
} from "./workspace/dashboards/layout";
export {
    isDashboardLayoutEmpty,
    IWidgetWithLayoutPath,
    LayoutPath,
    layoutWidgets,
    layoutWidgetsWithPaths,
    walkLayout,
} from "./workspace/dashboards/utils";
export {
    IWidget,
    IWidgetDefinition,
    IInsightWidget,
    IInsightWidgetBase,
    IInsightWidgetDefinition,
    IKpiWidget,
    IKpiWidgetBase,
    IKpiWidgetDefinition,
    WidgetType,
    isWidget,
    isWidgetDefinition,
    isInsightWidget,
    isInsightWidgetDefinition,
    isKpiWidget,
    isKpiWidgetDefinition,
    IWidgetReferences,
    SupportedWidgetReferenceTypes,
    widgetUri,
    widgetId,
    widgetType,
    IWidgetBase,
} from "./workspace/dashboards/widget";
export {
    ILegacyKpi,
    ILegacyKpiBase,
    ILegacyKpiComparisonDirection,
    ILegacyKpiComparisonTypeComparison,
    ILegacyKpiWithPopComparison,
    ILegacyKpiWithPreviousPeriodComparison,
    ILegacyKpiWithComparison,
    ILegacyKpiWithoutComparison,
    isLegacyKpiWithComparison,
    isLegacyKpiWithoutComparison,
    isLegacyKpi,
} from "./workspace/dashboards/kpi";
export {
    IWidgetAlertBase,
    IWidgetAlert,
    IWidgetAlertDefinition,
    IWidgetAlertCount,
    isWidgetAlert,
    isWidgetAlertDefinition,
} from "./workspace/dashboards/alert";
export {
    IDashboardAttachment,
    IScheduledMail,
    IScheduledMailDefinition,
    ScheduledMailAttachment,
    IScheduledMailBase,
} from "./workspace/dashboards/scheduledMail";
export { IWorkspaceUser, IWorkspaceUsersQuery, IWorkspaceUsersQueryOptions } from "./workspace/users";
export { IDateFilterConfigsQuery, IDateFilterConfigsQueryResult } from "./workspace/dateFilterConfigs";
export {
    IDateFilterConfig,
    AbsoluteFormType,
    AbsolutePresetType,
    AllTimeType,
    DateFilterGranularity,
    DateString,
    IAbsoluteDateFilterForm,
    IAbsoluteDateFilterPreset,
    IAllTimeDateFilterOption,
    IDateFilterOption,
    IRelativeDateFilterForm,
    IRelativeDateFilterPreset,
    IRelativeDateFilterPresetOfGranularity,
    OptionType,
    RelativeFormType,
    RelativeGranularityOffset,
    RelativePresetType,
    isAbsoluteDateFilterForm,
    isAbsoluteDateFilterPreset,
    isAllTimeDateFilterOption,
    isRelativeDateFilterForm,
    isRelativeDateFilterPreset,
    isDateFilterGranularity,
} from "./workspace/dateFilterConfigs/types";

export {
    CatalogItemType,
    CatalogItem,
    ICatalogGroup,
    ICatalogAttribute,
    ICatalogFact,
    ICatalogMeasure,
    ICatalogDateDataset,
    ICatalogDateAttribute,
    isCatalogAttribute,
    isCatalogFact,
    isCatalogMeasure,
    isCatalogDateDataset,
    ICatalogItemBase,
    IGroupableCatalogItemBase,
    GroupableCatalogItem,
    catalogItemMetadataObject,
} from "./workspace/fromModel/ldm/catalog";

export {
    IAttributeDisplayFormMetadataObject,
    isAttributeDisplayFormMetadataObject,
    IAttributeMetadataObject,
    isAttributeMetadataObject,
    IDataSetMetadataObject,
    isDataSetMetadataObject,
    IVariableMetadataObject,
    isVariableMetadataObject,
    IFactMetadataObject,
    isFactMetadataObject,
    IMetadataObjectDefinition,
    IMeasureMetadataObject,
    IMeasureMetadataObjectBase,
    isMeasureMetadataObject,
    IMeasureMetadataObjectDefinition,
    isMeasureMetadataObjectDefinition,
    IMetadataObject,
    IMetadataObjectBase,
    IMetadataObjectIdentity,
    isMetadataObject,
    MetadataObject,
    metadataObjectId,
    IDashboardMetadataObject,
    isDashboardMetadataObject,
} from "./workspace/fromModel/ldm/metadata";

export {
    DataColumnType,
    DatasetLoadStatus,
    IDataColumn,
    IDataHeader,
    IDatasetLoadInfo,
    IDatasetUser,
    IDataset,
} from "./workspace/fromModel/ldm/datasets";

export { IAttributeElement } from "./workspace/fromModel/ldm/attributeElement";

export {
    IMeasureExpressionToken,
    IObjectExpressionToken,
    IAttributeElementExpressionToken,
    ITextExpressionToken,
} from "./workspace/fromModel/ldm/measure";

export { IOrganization, IOrganizations, IOrganizationDescriptor } from "./organization";
export { ISecuritySettingsService, ValidationContext } from "./organization/securitySettings";
