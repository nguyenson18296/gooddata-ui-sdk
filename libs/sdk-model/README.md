# SDK domain models

This package provides domain models for GoodData.UI.

These domain models are backend-agnostic. This makes them reusable across different Analytical Backend implementations.
The package includes TypeScript type definitions, factory functions, functions to get or set certain properties
of the objects in an immutable way, and more. These are used in both the `@gooddata/sdk-backend-*` and `@gooddata/sdk-ui* packages`.

## License

(C) 2017-2021 GoodData Corporation

This project is under MIT License. See [LICENSE](https://github.com/gooddata/gooddata-ui-sdk/blob/master/libs/sdk-model/LICENSE).
