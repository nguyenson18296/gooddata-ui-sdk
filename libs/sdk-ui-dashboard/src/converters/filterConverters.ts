// (C) 2020-2021 GoodData Corporation
import {
    FilterContextItem,
    IFilterContext,
    IFilterContextDefinition,
    isDashboardAttributeFilter,
    ITempFilterContext,
    IWidgetDefinition,
} from "@gooddata/sdk-backend-spi";
import {
    newNegativeAttributeFilter,
    newPositiveAttributeFilter,
    newRelativeDateFilter,
    newAbsoluteDateFilter,
} from "@gooddata/sdk-model";
import isString from "lodash/isString";
import { IDashboardFilter } from "../types";

/**
 * Gets {@link IDashboardFilter} items for filters specified in given filterContext in relation to the given widget.
 *
 * @param filterContext - filter context to get filters for
 * @param widget - widget to use to get dateDataSet for date filters
 * @internal
 */
export function filterContextToFiltersForWidget(
    filterContext: IFilterContextDefinition | IFilterContext | ITempFilterContext | undefined,
    widget: IWidgetDefinition,
): IDashboardFilter[] {
    if (!filterContext) {
        return [];
    }

    return filterContextItemsToFiltersForWidget(filterContext.filters, widget);
}

/**
 * Gets {@link IDashboardFilter} items for filters specified as {@link FilterContextItem} instances.
 *
 * @param filterContextItems - filter context items to get filters for
 * @param widget - widget to use to get dateDataSet for date filters
 * @internal
 */
export function filterContextItemsToFiltersForWidget(
    filterContextItems: FilterContextItem[],
    widget: IWidgetDefinition,
): IDashboardFilter[] {
    return filterContextItems.map((filter) => {
        if (isDashboardAttributeFilter(filter)) {
            if (filter.attributeFilter.negativeSelection) {
                return newNegativeAttributeFilter(
                    filter.attributeFilter.displayForm,
                    filter.attributeFilter.attributeElements,
                );
            } else {
                return newPositiveAttributeFilter(
                    filter.attributeFilter.displayForm,
                    filter.attributeFilter.attributeElements,
                );
            }
        } else {
            if (filter.dateFilter.type === "relative") {
                return newRelativeDateFilter(
                    widget.dateDataSet!,
                    filter.dateFilter.granularity,
                    numberOrStringToNumber(filter.dateFilter.from!),
                    numberOrStringToNumber(filter.dateFilter.to!),
                );
            } else {
                return newAbsoluteDateFilter(
                    widget.dateDataSet!,
                    filter.dateFilter.from!.toString(),
                    filter.dateFilter.to!.toString(),
                );
            }
        }
    });
}

function numberOrStringToNumber(input: number | string): number {
    return isString(input) ? Number.parseInt(input) : input;
}
