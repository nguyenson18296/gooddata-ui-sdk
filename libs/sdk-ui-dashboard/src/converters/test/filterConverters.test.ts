// (C) 2020-2021 GoodData Corporation
import { ReferenceLdm } from "@gooddata/reference-workspace";
import { FilterContextItem, IFilterContext, IWidgetDefinition } from "@gooddata/sdk-backend-spi";
import { idRef } from "@gooddata/sdk-model";

import { filterContextToFiltersForWidget } from "../filterConverters";

describe("filterContextToFiltersForWidget", () => {
    const widget: IWidgetDefinition = {
        description: "",
        drills: [],
        ignoreDashboardFilters: [],
        title: "Widget",
        type: "insight",
        dateDataSet: ReferenceLdm.DateDatasets.Activity.ref,
        insight: idRef(ReferenceLdm.Insights.Test),
    };

    const getFilterContext = (filters: FilterContextItem[]): IFilterContext => ({
        description: "",
        filters,
        identifier: "foo",
        ref: idRef("foo"),
        title: "",
        uri: "/gdc/md/foo",
    });

    type Scenario = [string, IFilterContext | undefined];

    const scenarios: Scenario[] = [
        ["undefined filterContext", undefined],
        [
            "filterContext with positive attribute filter with values",
            getFilterContext([
                {
                    attributeFilter: {
                        attributeElements: { values: ["foo", "bar"] },
                        displayForm: ReferenceLdm.Account.Name.attribute.displayForm,
                        negativeSelection: false,
                    },
                },
            ]),
        ],
        [
            "filterContext with negative attribute filter with uris",
            getFilterContext([
                {
                    attributeFilter: {
                        attributeElements: { uris: ["foo", "bar"] },
                        displayForm: ReferenceLdm.Account.Name.attribute.displayForm,
                        negativeSelection: true,
                    },
                },
            ]),
        ],
        [
            "filterContext with relative date filter",
            getFilterContext([
                {
                    dateFilter: {
                        granularity: "GDC.time.date",
                        type: "relative",
                        from: -4,
                        to: -3,
                    },
                },
            ]),
        ],
        [
            "filterContext with absolute date filter",
            getFilterContext([
                {
                    dateFilter: {
                        granularity: "GDC.time.date",
                        type: "absolute",
                        from: "2020-01-01",
                        to: "2021-01-01",
                    },
                },
            ]),
        ],
    ];

    it.each(scenarios)("should handle %s", (_, filterContext) => {
        expect(filterContextToFiltersForWidget(filterContext, widget)).toMatchSnapshot();
    });
});
