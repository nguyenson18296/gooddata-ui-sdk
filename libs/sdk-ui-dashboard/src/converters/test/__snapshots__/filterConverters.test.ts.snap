// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filterContextToFiltersForWidget should handle filterContext with absolute date filter 1`] = `
Array [
  Object {
    "absoluteDateFilter": Object {
      "dataSet": Object {
        "identifier": "activity.dataset.dt",
        "type": "dataSet",
      },
      "from": "2020-01-01",
      "to": "2021-01-01",
    },
  },
]
`;

exports[`filterContextToFiltersForWidget should handle filterContext with negative attribute filter with uris 1`] = `
Array [
  Object {
    "negativeAttributeFilter": Object {
      "displayForm": Object {
        "identifier": "label.account.id.name",
        "type": "displayForm",
      },
      "notIn": Object {
        "uris": Array [
          "foo",
          "bar",
        ],
      },
    },
  },
]
`;

exports[`filterContextToFiltersForWidget should handle filterContext with positive attribute filter with values 1`] = `
Array [
  Object {
    "positiveAttributeFilter": Object {
      "displayForm": Object {
        "identifier": "label.account.id.name",
        "type": "displayForm",
      },
      "in": Object {
        "values": Array [
          "foo",
          "bar",
        ],
      },
    },
  },
]
`;

exports[`filterContextToFiltersForWidget should handle filterContext with relative date filter 1`] = `
Array [
  Object {
    "relativeDateFilter": Object {
      "dataSet": Object {
        "identifier": "activity.dataset.dt",
        "type": "dataSet",
      },
      "from": -4,
      "granularity": "GDC.time.date",
      "to": -3,
    },
  },
]
`;

exports[`filterContextToFiltersForWidget should handle undefined filterContext 1`] = `Array []`;
